packages/client/src/links/wsLink/wsClient/wsConnection.ts

```diff
@@ -0,0 +1,229 @@
+import { behaviorSubject } from '@trpc/server/observable';
+import { withResolvers } from './utils';
+
+/**
+ * Opens a WebSocket connection asynchronously and returns a promise
+ * that resolves when the connection is successfully established.
+ * The promise rejects if an error occurs during the connection attempt.
+ */
+function asyncWsOpen(ws: WebSocket) {
+  const { promise, resolve, reject } = withResolvers<void>();
+
+  ws.addEventListener('open', () => {
+    ws.removeEventListener('error', reject);
+    resolve();
+  });
+  ws.addEventListener('error', reject);
+
+  return promise;
+}
+
+interface PingPongOptions {
+  /**
+   * The interval (in milliseconds) between "PING" messages.
+   */
+  intervalMs: number;
+
+  /**
+   * The timeout (in milliseconds) to wait for a "PONG" response before closing the connection.
+   */
+  pongTimeoutMs: number;
+}
+
+/**
+ * Sets up a periodic ping-pong mechanism to keep the WebSocket connection alive.
+ *
+ * - Sends "PING" messages at regular intervals defined by `intervalMs`.
+ * - If a "PONG" response is not received within the `pongTimeoutMs`, the WebSocket is closed.
+ * - The ping timer resets upon receiving any message to maintain activity.
+ * - Automatically starts the ping process when the WebSocket connection is opened.
+ * - Cleans up timers when the WebSocket is closed.
+ *
+ * @param ws - The WebSocket instance to manage.
+ * @param options - Configuration options for ping-pong intervals and timeouts.
+ */
+function setupPingInterval(
+  ws: WebSocket,
+  { intervalMs, pongTimeoutMs }: PingPongOptions,
+) {
+  let pingTimeout: ReturnType<typeof setTimeout> | undefined;
+  let pongTimeout: ReturnType<typeof setTimeout> | undefined;
+
+  function start() {
+    pingTimeout = setTimeout(() => {
+      ws.send('PING');
+      pongTimeout = setTimeout(() => {
+        ws.close();
+      }, pongTimeoutMs);
+    }, intervalMs);
+  }
+
+  function reset() {
+    clearTimeout(pingTimeout);
+    start();
+  }
+
+  function pong() {
+    clearTimeout(pongTimeout);
+    reset();
+  }
+
+  ws.addEventListener('open', start);
+  ws.addEventListener('message', ({ data }) => {
+    clearTimeout(pingTimeout);
+    start();
+
+    if (data === 'PONG') {
+      pong();
+    }
+  });
+  ws.addEventListener('close', () => {
+    clearTimeout(pingTimeout);
+    clearTimeout(pongTimeout);
+  });
+}
+
+export interface WebSocketConnectionOptions {
+  WebSocketPonyfill?: typeof WebSocket;
+  promiseUrl: Promise<string>;
+  keepAlive: PingPongOptions & {
+    enabled: boolean;
+  };
+}
+
+/**
+ * Manages a WebSocket connection with support for reconnection, keep-alive mechanisms,
+ * and observable state tracking.
+ */
+export class WsConnection {
+  static connectCount = 0;
+  public id = ++WsConnection.connectCount;
+
+  private readonly WebSocketPonyfill: typeof WebSocket;
+  private readonly promiseUrl: Promise<string>;
```

Comment:
Previously, the `url`-callback would be invoked for every reopen so that people could dynamically switch server/endpoint

We should store the function that generates the promise here rather than the promise